### DISCUSSION TITLE
How to throw an exception inside the Unirest 3.14.5 thenConsume method?

### DISCUSSION THREAD

**Discussion by zhanyeye** [Kong/unirest-java#517] [Q&A] [answered]
Hi,

I'm using Unirest `3.14.5` for HTTP requests and handling responses with `.thenConsume(res -> {})`. 
I want to throw an exception if the response status isn't 200 or if there's an issue writing the response to a file. 
However, exceptions don't seem to propagate as expected.

Here's the simplified code:

```java
public void downloadLargeJson(Params params, File targetFile) {
    Unirest.post(URL)
            .body(new Gson().toJson(params))
            .thenConsume(response -> {
                if (response.getStatus() != 200) {
                    log.error("http code: {}, response: {}", response.getStatus(), response.getContentAsString());
                    throw new RuntimeException("request fail");
                }
                try (InputStream responseStream = response.getContent()) {
                    writeInputStreamToFile(responseStream, targetFile);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
}
```

And here's how I've set up the retry logic:
```java
public void downloadLargeJsonWithRetry(Params params, File targetFile) {
    RetryPolicy<Object> retryPolicy = RetryPolicy.builder()
            .handle(RuntimeException.class)
            .withDelay(Duration.ofSeconds(10))
            .withMaxRetries(5)
            .build();

    Failsafe.with(retryPolicy).run(() -> downloadLargeJson(params, targetFile));
}
```
The retry logic is intended to handle failed requests, but it doesn't seem to activate upon exceptions thrown within .thenConsume. Could you provide guidance on ensuring that the retry logic works as intended for both HTTP errors and file writing issues?

Thanks in advance for your assistance.

**Comment by ryber** [comment]
What version are you on?

**Reply by zhanyeye** [reply]
3.14.5

**Reply by ryber** [reply, marked as answer]
This was "fixed" in the 4.x line. see https://github.com/Kong/unirest-java/issues/506
There are no plans to backport this into the 3.x line which is only being maintained for security issues at this point
